pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,270 @@
+# Copyright 2017 Pierre-Étienne Meunier
+#
+# This file is licensed under the Apache-2.0 license, and under the
+# MIT license, at your convenience.
+
+{ lib, buildPlatform, stdenv }:
+
+let buildCrate = { crateName, crateVersion, dependencies, complete, crateFeatures, libName, build, release, libPath, crateType, metadata, crateBin, finalBins, verboseBuild }:
+
+      let depsDir = builtins.foldl' (deps: dep: deps + " " + dep.out) "" dependencies;
+          completeDepsDir = builtins.foldl' (deps: dep: deps + " " + dep.out) "" complete;
+          deps =
+            builtins.foldl' (deps: dep:
+              let extern = lib.strings.replaceStrings ["-"] ["_"] dep.libName; in
+              deps + (if dep.crateType == "lib" then
+                 " --extern ${extern}=${dep.out}/lib${extern}-${dep.metadata}.rlib"
+              else
+                 " --extern ${extern}=${dep.out}/lib${extern}-${dep.metadata}.so")
+            ) "" dependencies;
+          optLevel = if release then 3 else 0;
+          rustcOpts = (if release then "-C opt-level=3" else "-C debuginfo=2");
+          rustcMeta = "-C metadata=" + metadata + " -C extra-filename=-" + metadata;
+      in ''
+
+      norm="$(printf '\033[0m')" #returns to "normal"
+      bold="$(printf '\033[0;1m')" #set bold
+      green="$(printf '\033[0;32m')" #set red
+      boldgreen="$(printf '\033[0;1;32m')" #set bold, and set red.
+
+      mkdir -p target/deps
+      mkdir -p target/build
+      chmod uga+w target -R
+      for i in ${completeDepsDir}; do
+         ln -s -f $i/*.rlib target/deps #*/
+         ln -s -f $i/*.so target/deps #*/
+         if [ -e "$i/link" ]; then
+            cat $i/link >> target/link
+            cat $i/link >> target/link.final
+         fi
+         if [ -e $i/env ]; then
+            source $i/env
+         fi
+      done
+      if [ -e target/link ]; then
+        sort -u target/link > target/link.sorted
+        mv target/link.sorted target/link
+        sort -u target/link.final > target/link.final.sorted
+        mv target/link.final.sorted target/link.final
+        tr '\n' ' ' < target/link > target/link_
+      fi
+      EXTRA_BUILD=""
+      BUILD_OUT_DIR=""
+      export CARGO_PKG_NAME=${crateName}
+      export CARGO_PKG_VERSION=${crateVersion}
+      export CARGO_CFG_TARGET_ARCH=$(echo ${buildPlatform.system} | sed -e "s/\([^-]*\)-\([^-]*\)/\1/")
+      export CARGO_CFG_TARGET_OS=$(echo ${buildPlatform.system} | sed -e "s/\([^-]*\)-\([^-]*\)/\2/")
```

Comment:
buildPlatform.parsed.kernel.name

Reply:
Where is the documentation for this? If it doesn't exist, I'd like to add it somewhere.

Reply:
I found out, while playing the `nix-repl`:

```console
$ nix-repl
Welcome to Nix version 1.11.15. Type :? for help.
nix-repl> :l <nixpkgs>
Added 7885 variables.
nix-repl> buildPlatform.parsed
{ _type = "system"; abi = { ... }; cpu = { ... }; kernel = { ... }; vendor = { ... }; }
nix-repl>buildPlatform.parsed.kernel
{ _type = "kernel"; execFormat = { ... }; families = { ... }; name = "linux"; }
```

I think this was merged recently for cross compiling. It comes from `lib/systems`.

Reply:
This might belong to the nixpkgs documentation then, located in `doc`